File: Data structures/linked list/swapnode.cpp
#include <iostream>
using namespace std;
class Node
{
public:
    int data;
    Node *next;
};
void push(Node **head, int data)
{
    Node *n = new Node();
    n->data = data;
    n->next = *head;
    *head = n;
}
void print(Node *head)
{
    Node *d = head;
    while (d)
    {
        cout << d->data << '\n';
        d = d->next;
    }
}
void swap(Node **head, int x, int y)
{
    Node *prevx = NULL, *prevy = NULL, *nodex = *head, *nodey = *head, *temp;
    for (int i = 1; i < x; i++)
    {
        prevx = nodex;
        nodex = nodex->next;
    }
    for (int i = 0; i < y - 1; i++)
    {
        prevy = nodey;
        nodey = nodey->next;
    }
    prevx->next = nodey;
    prevy->next = nodex;
    temp = nodex->next;
    nodex->next = nodey->next;
    nodey->next = temp;
}
int main()
{
    Node *head = NULL;
    push(&head, 7);
    push(&head, 6);
    push(&head, 5);
    push(&head, 4);
    push(&head, 3);
    push(&head, 2);
    push(&head, 1);
    print(head);
    swap(&head, 2, 6);
    cout<<"----------------\n";
    print(head);
    return 0;
}

File: Data structures/linked list/swapNode.java
public class swapNode {
    class Node {
        int data;
        Node next;

        Node(int data) {
            this.data = data;
        }
    }

    void insertNode(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            newNode.next = head;
            head = newNode;
        }
    }

    void swap(int x, int y) {
        Node prevX = null;
        Node currX = head;
        while (currX != null && currX.data != x) {
            prevX = currX;
            currX = currX.next;
        }
        Node prevY = null;
        Node currY = head;
        while (currY != null && currY.data != y) {
            prevY = currY;
            currY = currY.next;
        }
        if (currY == null || currX == null)
            return;
        if (currX == currY)
            return;
        if (currX == head) {
            currY.next = head.next;
            head = currY;
        } 
        else {
            prevX.next = currY;
        }
        if (currY == head) {
            currX.next = head.next;
            head = currX;
        } 
        else {
            prevY.next = currX;
        }
        Node temp = currX.next;
        currX.next = currY.next;
        currY.next = temp;
    }

    void printList() {
        Node trav = head;
        while (trav != null) {
            System.out.print(trav.data + " ");
            trav = trav.next;
        }
        System.out.println();
    }

    static Node head;

    public static void main(String args[]) {
        swapNode sn = new swapNode();
        for (int i = 0; i < 10; i++) {
            sn.insertNode(i);
        }
        sn.printList();
        sn.swap(9, 6);
        sn.printList();
    }
}